Please **submit this Jupyter notebook through Canvas** no later than **Monday November 4, 8:59**, before the start of the lecture.

Homework is in **groups of two**, and you are expected to hand in original work. Work that is copied from another group will not be accepted.

In [3]:
import numpy as np
import time

-----
# Exercise 0
In the box below, write down the names + student ID of the people in your group.

Name1: Wiebe Jelsma <br>
Name2: Julien Fer <br>
Student_number1: 12468223 <br>
Student_number2: 10649441 <br>

-----
# Exercise 1
This exercise is a variant of exercise 1.6 in the book.

## (a) 
Lookup the Taylor series for $\cos(x)$.&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _<font size=2>(You don't have to hand in the series expansion)</font>_
## (b)
What are the forward and backward errors if we approximate $\cos(x)$ by the first **two** nonzero terms in the Taylor series at $x = 0.2$, $x = 1.0$ and $x = 2.0$?

In [4]:
def two_non_zero(x):
    y = 1-(x**2/2)
    return y

x1 = 0.2
x2 = 1
x3 = 2

FE_0 = two_non_zero(x1) - np.cos(0.2)
FE_1 = two_non_zero(x2) - np.cos(1)
FE_2 = two_non_zero(x3) - np.cos(2)
print("Forward error of x = 0.2 is ", FE_0)
print("Forward error of x = 1 is ", FE_1)
print("Forward error of x = 2 is ", FE_2)

BE_0 = np.arccos(two_non_zero(x1)) - x1
BE_1 = np.arccos(two_non_zero(x2)) - x2
BE_2 = np.arccos(two_non_zero(x3)) - x3
print("Backward error of x = 0.2 is ", BE_0)
print("Backward error of x = 1 is ", BE_1)
print("Backward error of x = 2 is ", BE_2)

Forward error of x = 0.2 is  -6.657784124164401e-05
Forward error of x = 1 is  -0.040302305868139765
Forward error of x = 2 is  -0.5838531634528576
Backward error of x = 0.2 is  0.00033484232311967177
Backward error of x = 1 is  0.04719755119659785
Backward error of x = 2 is  1.1415926535897931


Write your answer, using $\LaTeX$, in this box. <br>
Pay attention, there might be rounding errors! We choose to display the first 4 digits, because the book chose the same.
We made use of the Maclaurin series, so we assume a = 0. 
### Taylor series:

Taylor series for cos(x):
$1-\frac{x^2}{2!} + \frac{x^4}{4!} - ... + \frac{x^n}{n!}$
For the first two nonzero terms we use $1-\frac{x^2}{2!}$ <br>
When x = 0.2, this will give us: <br>
$1 - \frac{0.2^2}{2}$ which results in 0.98. <br>
When x is 1, this gives $1 - \frac{1^2}{2}$ which results in 0.5 and x = 2 gives us 
$1 - \frac{2^2}{2}$ which results in -1.<br>
### Forward erros:

The real value of $cos(0.2) = 9.8001e^{-1}$, $cos(1) = 5.4030e^{-1}$ and $cos(2) = -4.1615e^{-1}$. 
The discrepancy between the computed and true values, $\Delta y = \hat{y} - y$, is called the forward error. 

For $x = 0.2$, we see that the forward error is $0.98 - cos(0.2) = -6.6578e^{-5}$.<br>
For $x = 1$ we will see a forward error of $0.5 - cos(1) = -4.0302e^{-2}$. <br>
$x = 2$ we find $-1 - cos(2) = -5.8385e^{-1}$. We see that the forward error gets bigger when x becomes bigger. <br> 
### Backward errors:

We can find the backwards error with arccos. You have to take the arcoss of the value you found. For $x = 1:$
we take the arccos of the value we found. We filled in $1 - \frac{1^2}{2}$ which leads to 0.5. We then take the arccos of 0.5, which is 1.0472. We then subtract x, so 1. We find $arccos(0.5) - 1 = 4.7198e^{-2}$. <br>
For $x = 0.2$, we found $1 - \frac{0.2^2}{2}$ which is 0.98. We take arccos of 0.98, which is 0.2003. We find $arccos(0.98) - 0.2 = 3.3484e^{-4}$. <rb>
Now $x = 2$. We found that value -1. Arccos of -1 is 3.1416, thus the backward error of $x = 2$ is calculated by $arccos(-1) - 2 = 1.1416$.
<br>
### Summary:
x | forward error | backward error
:-- | :-: | --:
0.2 | -6.6578$e^{-5}$ | 3.3484$e^{-4}$
1 | -4.0302$e^{-2}$ | 4.7198$e^{-2}$
2 | -5.8385$e^{-1}$ | 1.1416

## (c)
What are the forward and backward errors if we approximate $\cos(x)$ by the first **three** nonzero terms in the Taylor series at $x = 0.2$, $x = 1.0$ and $x = 2.0$?

In [5]:
def three_non_zero(x):
    y = 1-(x**2/2)+(x**4/(24))
    return y
FE_0_3 = three_non_zero(x1) - np.cos(0.2)
FE_1_3 = three_non_zero(x2) - np.cos(1)
FE_2_3 = three_non_zero(x3) - np.cos(2)
print("Forward error of x = 0.2 is ", FE_0_3)
print("Forward error of x = 1 is ", FE_1_3)
print("Forward error of x = 2 is ", FE_2_3)

BE_0_3 = np.arccos(three_non_zero(x1)) - x1
BE_1_3 = np.arccos(three_non_zero(x2)) - x2
BE_2_3 = np.arccos(three_non_zero(x3)) - x3
print("Backward error of x = 0.2 is ", BE_0_3)
print("Backward error of x = 1 is ", BE_1_3)
print("Backward error of x = 2 is ", BE_2_3)

Forward error of x = 0.2 is  8.882542501531532e-08
Forward error of x = 1 is  0.0013643607985268646
Forward error of x = 2 is  0.08281350321380904
Backward error of x = 0.2 is  -4.4710234142764094e-07
Backward error of x = 1 is  -0.0016222452979235413
Backward error of x = 2 is  -0.0893667637509814


Write your answer, using $\LaTeX$, in this box.
### Taylor series:
Taylor series for cos(x) with three nonzero is: <br>
$1-\frac{x^2}{2!} + \frac{x^4}{4!}$ <br>
When x = 0.2 we get $1-\frac{0.2^2}{2!} + \frac{0.2^4}{4!}$ we get $9.8007e^{-1}$. <br>
For x = 1 we find $5.4167e^{-1}$ and for x = 2 we get $-3.3333e^{-1}$ <br>
### Forward error: 

$x = 0.2: $ will lead to $9.8007e^{-1} - cos(0.2) = 8.8819e^{-8}$, <br> $x = 1$ gives $5.4167e^{-1} - cos(1) = 1.3644e^{-3}$ and <br>$x = 2$ results in $-\frac{1}{3}$ $- cos(2) = 8.2814e^{-2}$. <br>Just like we saw in 1.B, the forward errors become lower when x becomes lower as well. <br>
### Backward error
$x = 0.2:$ <br>
The value we found was $9.8007e^{-1}$, we use this for arccos.  <br>
$arccos(9.8007e^{-1}) - 0.2 =  -1.683e^{-4}$
<br>
$x = 1:$ the value we found was $5.4167e^{-1}$, we use this for arccos. This gives: 0.9983381. <br>
$arccos(5.4167e^{-1}) - 1 = -1.6222e^{-3}$
<br>
$x = 2:$ the value we found was -1/3, we use this for arccos. This gives: 1.91063324. <br>
$arccos(-1/3) - 2 = -8.9367e^{-2}$


### Summary:
x | forward error | backward error
:-- | :-: | --:
0.2 | 8.8819$e^{-8}$ | -4.47$e^{-7}$
1 | 1.3644$e^{-3}$ | -1.6222$e^{-3}$
2 | 8.2814$e^{-2}$ | -8.9367$e^{-2}$

## (d)
Compute the relative condition of $x \mapsto \cos(x)$ at $x = 0.2$, $x=1.0$ and $x=2.0$.

### For 2 non-zero:
(forward error/y) / (backward error / x)
##### $x = 0.2$:
$ \frac{\frac{-6.6578e^{-5}}{cos(0.2)}} {\frac{3.3484e^{-4}}{0.2}} = -4.0576e^{-2}$
##### $x = 1$:
$ \frac{\frac{-4.0302e^{-2}}{cos(1)}} {\frac{4.7198e^{-2}}{1}} = -1.5804$
##### $x = 2$:

$ \frac{\frac{-5.8385e^{-1}}{cos(2)}} {\frac{1.1416}{2}} = 2.4579$

### For 3 non-zero:
##### $x = 0.2$:
$\frac{\frac{8.8819e^{-8}}{cos(0.2)}} {\frac{-4.47e^{-7}}{0.2}} = -4.0548e^{-2}$
##### $x = 1$:
$\frac{\frac{1.3644e^{-3}}{cos(1)}} {\frac{-1.6222e^{-3}}{1}} = -1.5567$
##### $x = 2$:
$\frac{\frac{8.2814e^{-2}}{cos(2)}} {\frac{-8.9367e^{-2}}{2}} = 4.4536$

### Summary:
x | 2 non-zero | 3 non-zero
:-- | :-: | --:
0.2 | $-4.0576e^{-2}$ | $-4.0548e^{-2}$
1 | -1.5804 | -1.5567
2 | 2.4579 | 4.4536

-----
# Exercise 2
This is computer exercise 1.16 from the book.

Write a program that sums $n$ random, **single-precision** floating-point numbers $x_i$, uniformly distributed on the interval $[0,1]$. Sum the numbers in each of the following ways (use only single-precision floating-point variables unless specifically indicated otherwise).

## (a)
Sum the numbers in the order in which they were generated, using a double-precision variable in which to accumulate the sum.

Put your solution code in the box below.

In [9]:
def sum_double_precision_normal(numbers):
    """
    For a given amount (n) of numbers random single-precision floating-point numbers are generated and summed. 
    The summation is in double-precision and in the order of the genreated numbers.
    """
    
    total = np.float64(0.0)
    for number in numbers:
        total += number
        
    return total

## (b) 
Sum the numbers in the order in which they were generated, this time using a single-precision accumulator.

In [10]:
def sum_single_precision_normal(numbers):
    """
    For a given amount (n) of numbers random single-precision floating-point numbers are generated and summed. 
    The summation is in single-precision and in the order of the genreated numbers.
    """
    
    total = np.float32(0.0)
    for number in numbers:
        total += number
        
    return total

## (c)
Use the following _compensated summation_ algorithm (due to Kahan), again using only single precision, to sum the numbers in the order in which they were generated:

$
s = x_1\\
c = 0\\
\textbf{for } i = 2 \textbf{ to } n\\
\quad y = x_i - c\\
\quad t = s+y\\
\quad c = (t-s) - y\\
\quad s = t\\
\textbf{end}
$

In [11]:
def comp_sum_normal(numbers):
    """
    Compensated summation algorithm (Kahan) in the order in which they were genretaded
    """
    
    total = numbers[0]
    c = np.float32(0.0)
    
    for i in range(1, len(numbers)):
        y = numbers[i] - c
        t = total + y
        c = (t - total) - y
        total = t
    
    return total

## (d) 
Sum the numbers in order of increasing magnitude (this will require that the numbers be sorted
before summing, for which you may use a library sorting routine).

In [12]:
def sum_single_precision_increasing(numbers):
    """
    For a given amount (n) of numbers random single-precision floating-point numbers are generated and summed. 
    The summation is in single-precision and in increasing order.
    """
    
    sorted_numbers = np.sort(numbers)
    total = np.float32(0.0)
    for number in sorted_numbers:
        total += number
        
    return total

def comp_sum_increasing(numbers):
    """
    Compensated summation algorithm (Kahan) in the increasing order.
    """
    
    sorted_numbers = np.sort(numbers)
    total = sorted_numbers[0]
    c = np.float32(0.0)
    
    for i in range(1, len(sorted_numbers)):
        y = sorted_numbers[i] - c
        t = total + y
        c = (t - total) - y
        total = t
    
    return total

## (e) 
Sum the numbers in order of decreasing magnitude (i.e., reverse the order of summation from part (d)).

In [13]:
def sum_single_precision_decreasing(numbers):
    """
    For a given amount (n) of numbers random single-precision floating-point numbers are generated and summed. 
    The summation is in single-precision and in decreasing order.
    """
    
    sorted_numbers = np.sort(numbers)[::-1]
    total = np.float32(0.0)
    for number in sorted_numbers:
        total += number
        
    return total

def comp_sum_decreasing(numbers):
    """
    Compensated summation algorithm (Kahan) in the decreasing order.
    """
    
    sorted_numbers = np.sort(numbers)[::-1]
    total = sorted_numbers[0]
    c = np.float32(0.0)
    
    for i in range(1, len(sorted_numbers)):
        y = sorted_numbers[i] - c
        t = total + y
        c = (t - total) - y
        total = t
    
    return total

## (f)

Run your program for various values of $n$ and compare the results for methods (a) through (e). You may need to use a fairly large value for $n$ to see a substantial difference. How do the methods rank in terms of accuracy, and why? How do the methods compare in cost? Can you explain why the algorithm in part (c) works?

Write your answer, using $\LaTeX$, in this box.

Tests are run for various values of $n:(10, 100, 1000, 10000, 100000, 1000000)$ (see code below). The differences with the summation found with the double-precision method of exercise a and the other methods are for small $n$ (less than 100) approximately the same, however, the compenated summation algorithm already manages to perform slightly more accurate that the other algorithms. Moreover, for a greater values of $n$ we note more significant effect:
* The summation algorithm of exercise c (the compensted summation algrotihm) is in all cases the most accurate (similar to the summation of exercise a). It is even insensitive to the order of summation (given, increasing or decreasing) and always yields the same result. This is not the case for the basic summation algorithm. For the basic summation algorithm it seems most accurate to perform the summation with an increasing order of numbers. The summation performed with the given order is more accurate than the algorithm with decreasing magnitude. So the complete rank would be: (1) the compensated summation algorithm (exercise c, given/increasing/decreasing order); (2) basic summation algorithm (increasing order); (3) basic summation algorithm (given order); and finally (4) basic summation algorithm (decreasing order).
* The compensated summation algorithm is the most time consuming of all. This makes sense as it contains the most elmentary operations of all. Furthermore, we note that the basic summation in increasing order is slower than performing the summation in decreasing order which in turn is slower than the basic summation in given order (might be to the fact that we need to sort the floating point numbers)

The algorithm in part (c) works because it makes use of two different accumulators: total holds the sum and c accumulates the part not assimilated into total, to nudge the low-order part of total the next time around. Thus the summation proceeds with "guard digits" in c, which is better than not having any, but is not as good as performing the calculations with double the precision of the input.

In [14]:
N = [10 ** i for i in range(1, 7)]

for amount in N:
    values = np.random.uniform(0, np.nextafter(1, 2), amount).astype("float32")
    total1 = sum_double_precision_normal(values)
    
    start = time.time()
    total2 = sum_single_precision_normal(values)
    end2 = time.time() - start
    diff2 = total1 - total2
    
    start = time.time()
    total3 = comp_sum_normal(values)
    end3 = time.time() - start
    diff3 = total1 - total3
    
    start = time.time()
    total4 = sum_single_precision_increasing(values)
    end4 = time.time() - start
    diff4 = total1 - total4
    
    start = time.time()
    total5 = comp_sum_increasing(values)
    end5 = time.time() - start
    diff5 = total1 - total5
    
    start = time.time()
    total6 = sum_single_precision_decreasing(values)
    end6 = time.time() - start
    diff6 = total1 - total6
    
    start = time.time()
    total7 = comp_sum_decreasing(values)
    end7 = time.time() - start
    diff7 = total1 - total7
    
    print(f"For {amount} random numbers betweeen 0 and 1")
    print()
    print(f"Summation in given order (double precision): {total1}")
    print(f"Summation in given order (single precision): {total2} || Difference is: {diff2} || Time: {end2}")
    print(f"Compensated sumation in given order: {total3} || Difference is: {diff3} || Time: {end3}")
    print(f"Summation in increasing order (single precision): {total4} || Difference is: {diff4} || Time: {end4}")
    print(f"Compensated summation in increasing order: {total5} || Difference is: {diff5} || Time: {end5}")
    print(f"Summation in decreasing order (single precision): {total6} || Difference is: {diff6} || Time: {end6}")
    print(f"Compensated summation in decreasing order: {total7} || Difference is: {diff7} || Time: {end7}")
    print("=============================================================================")
    

For 10 random numbers betweeen 0 and 1

Summation in given order (double precision): 5.705262541770935
Summation in given order (single precision): 5.705262184143066 || Difference is: 3.5762786865234375e-07 || Time: 0.0
Compensated sumation in given order: 5.705262660980225 || Difference is: -1.1920928955078125e-07 || Time: 0.0
Summation in increasing order (single precision): 5.705262660980225 || Difference is: -1.1920928955078125e-07 || Time: 0.0
Compensated summation in increasing order: 5.705262660980225 || Difference is: -1.1920928955078125e-07 || Time: 0.0
Summation in decreasing order (single precision): 5.705262660980225 || Difference is: -1.1920928955078125e-07 || Time: 0.0
Compensated summation in decreasing order: 5.705262660980225 || Difference is: -1.1920928955078125e-07 || Time: 0.0
For 100 random numbers betweeen 0 and 1

Summation in given order (double precision): 46.159665250452235
Summation in given order (single precision): 46.15966796875 || Difference is: -2.718297